src/Microsoft.Android.Sdk.ILLink/TypeMappingStep.cs

```diff
@@ -35,55 +56,160 @@ protected override void ProcessAssembly (AssemblyDefinition assembly)
-		foreach (var (javaName, list) in TypeMappings) {
-			/*
-			 * IL_0000: ldarg.0
-			 * IL_0001: ldfld class [System.Runtime]System.Collections.Generic.IDictionary`2<string, class [System.Runtime]System.Type> Microsoft.Android.Runtime.NativeAotTypeManager::TypeMappings
-			 * IL_0006: ldstr "java/lang/Object"
-			 * IL_000b: ldtoken [Mono.Android]Java.Lang.Object
-			 * IL_0010: call class [System.Runtime]System.Type [System.Runtime]System.Type::GetTypeFromHandle(valuetype [System.Runtime]System.RuntimeTypeHandle)
-			 * IL_0015: callvirt instance void class [System.Runtime]System.Collections.Generic.IDictionary`2<string, class [System.Runtime]System.Type>::Add(!0, !1)
-			 */
-			il.Emit (Mono.Cecil.Cil.OpCodes.Ldarg_0);
-			il.Emit (Mono.Cecil.Cil.OpCodes.Ldfld, field);
-			il.Emit (Mono.Cecil.Cil.OpCodes.Ldstr, javaName);
-			il.Emit (Mono.Cecil.Cil.OpCodes.Ldtoken, module.ImportReference (SelectTypeDefinition (javaName, list)));
-			il.Emit (Mono.Cecil.Cil.OpCodes.Call, getTypeFromHandle);
-			il.Emit (Mono.Cecil.Cil.OpCodes.Callvirt, addMethod);
+			var il = method.Body.GetILProcessor ();
+
+			var orderedMapping = TypeMappings.Select(kvp => (Hash (kvp.Key), SelectTypeDefinition (kvp.Key, kvp.Value))).OrderBy (x => x.Item1);
+
+			var targets = new List<Instruction> ();
+			foreach (var name in names) {
+				targets.Add (il.Create (OpCodes.Ldstr, name));
+			}
+
+			il.Emit (OpCodes.Ldarg_0);
+			il.Emit (OpCodes.Switch, targets.ToArray ());
+
+			var defaultTarget = il.Create (OpCodes.Ldnull);
+			il.Emit (OpCodes.Br, defaultTarget);
+
+			foreach (var target in targets) {
+				il.Append (target);
+				il.Emit (OpCodes.Ret);
+			}
+
+			il.Append (defaultTarget);
+			il.Emit (OpCodes.Ret);
 		}
 
-		il.Emit (Mono.Cecil.Cil.OpCodes.Ret);
+		void GenerateHashes (ulong[] hashes)
+		{
+			// Sanity check: hashes must be unique and sorted
+			if (hashes.Length > 0) {
+				ulong previous = hashes[0];
+				for (int i = 1; i < hashes.Length; ++i) {
+					if (hashes[i] == previous) {
+						throw new InvalidOperationException ($"Duplicate hashes");
+					} else if (hashes[i] < previous) {
+						throw new InvalidOperationException ($"Hashes are not in ascending order");
+					}
+
+					previous = hashes[i];
+				}
+			}
+
+			// Create static array struct for `byte[#number-of-hashes]`
+			var arrayType = new TypeDefinition (
+				"",
+				"HashesArray",
+				TypeAttributes.NestedPrivate | TypeAttributes.ExplicitLayout,
+				module.ImportReference (typeof (ValueType)))
+			{
+				PackingSize = 1,
+				ClassSize = hashes.Length * sizeof (ulong),
+			};
+
+			type.NestedTypes.Add (arrayType);
+
+			// Create static field to store the raw bytes
+			var bytesField = new FieldDefinition ("s_hashes", FieldAttributes.Private | FieldAttributes.Static | FieldAttributes.InitOnly, arrayType);
+			bytesField.InitialValue = hashes.Select(h => BitConverter.GetBytes (h)).SelectMany (x => x).ToArray ();
```

Comment:
Are there endianness issues here? For example, will the host (IL generator) and target always have the same endianness?

Reply:
You're right. I need to revisit this.